pkg/libartifact/store/store.go

```diff
@@ -0,0 +1,375 @@
+		Versioned: specs.Versioned{SchemaVersion: 2},
+	}
+	rawData, err := json.Marshal(&index)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(as.indexPath(), rawData, 0o644)
+}
+
+func (as ArtifactStore) indexPath() string {
+	return filepath.Join(as.storePath, specV1.ImageIndexFile)
+}
+
+// getArtifacts returns an ArtifactList based on the artifact's store.  The return error and
+// unused opts is meant for future growth like filters, etc so the API does not change.
+func (as ArtifactStore) getArtifacts(ctx context.Context, _ *libartTypes.GetArtifactOptions) (libartifact.ArtifactList, error) {
+	var (
+		al libartifact.ArtifactList
+	)
+
+	lrs, err := layout.List(as.storePath)
+	if err != nil {
+		return nil, err
+	}
+	for _, l := range lrs {
+		imgSrc, err := l.Reference.NewImageSource(ctx, as.SystemContext)
+		if err != nil {
+			return nil, err
+		}
+		manifests, err := getManifests(ctx, imgSrc, nil)
+		imgSrc.Close()
+		if err != nil {
+			return nil, err
+		}
+		artifact := libartifact.Artifact{
+			Manifests: manifests,
+		}
+		if val, ok := l.ManifestDescriptor.Annotations[specV1.AnnotationRefName]; ok {
+			artifact.SetName(val)
+		}
+
+		al = append(al, &artifact)
+	}
+	return al, nil
+}
+
+// getManifests takes an imgSrc and starting digest (nil means "top") and collects all the manifests "under"
+// it.  this func calls itself recursively with a new startingDigest assuming that we are dealing with
+// an index list
+func getManifests(ctx context.Context, imgSrc types.ImageSource, startingDigest *digest.Digest) ([]manifest.OCI1, error) {
+	var (
+		manifests []manifest.OCI1
+	)
+	b, manifestType, err := imgSrc.GetManifest(ctx, startingDigest)
+	if err != nil {
+		return nil, err
+	}
+
+	// this assumes that there are only single, and multi-images
+	if !manifest.MIMETypeIsMultiImage(manifestType) {
+		// these are the keepers
+		mani, err := manifest.OCI1FromManifest(b)
+		if err != nil {
+			return nil, err
+		}
+		manifests = append(manifests, *mani)
+		return manifests, nil
+	}
+	// We are dealing with an oci index list
+	maniList, err := manifest.OCI1IndexFromManifest(b)
+	if err != nil {
+		return nil, err
+	}
+	for _, m := range maniList.Manifests {
+		iterManifests, err := getManifests(ctx, imgSrc, &m.Digest)
+		if err != nil {
+			return nil, err
+		}
+		manifests = append(manifests, iterManifests...)
+	}
+	return manifests, nil
+}
+
+func checkForEmptyStanzaFile(path string) error {
+	if _, err := os.Stat(path); err == nil {
+		return nil
+	}
+	return os.WriteFile(path, specV1.DescriptorEmptyJSON.Data, 0644)
+}
+
+func determineManifestType(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	// DetectContentType looks at the first 512 bytes
+	b := make([]byte, 512)
+	// Because DetectContentType will return a default value
+	// we don't sweat the error
+	n, _ := f.Read(b)
+	return http.DetectContentType(b[:n]), nil
```

Comment:
Does this still work if n is 0?

Reply:
Yes, it does, but I got "text/plain; charset=utf-8" when I tried it.  Running the value `DetectContentType()` returns through `mime.ParseMediaType()` to split the MIME type name from the parameters, and then using just the type name, would remove that second part.

Reply:
yup, it returns `text/plain` . see https://go.dev/play/p/2KZhZuHjoCT

Reply:
The code above seems to want to default to `application/octet-stream` which is not `text/plain; charset=utf-8`?!
Defaulting to text/plain seems wrong for the artifact use case I think??

Reply:
If anything, that's something to be sorted out in the standard library, where the godoc suggested to me that we should be getting "application/octet-string" back.  The detection function can't return an error, so unless we encounter an error reading the file (in which case the whole thing should probably fail), we can manage without deciding on a default value.